### ISSUE TITLE
Check integer types implement correct set of ops

### ISSUE THREAD

**Issue by tcharding** [rust-bitcoin/rust-bitcoin#1639]
`pow` types implement `Add` but not `AddAsign`.

We should check all the integer types and ensure they implement a sane set of `core::ops`. This is not as trivial as just implementing _all_ of them because the types often implement some kind of Bitcoin abstraction.

This is a good issue to do if you'd like to spend time thinking deeply about Bitcoin integer abstractions.

### TODO

To find a list of types to check consider using
```
git grep '\((u|i)(8|16|32|64)\)' | grep -v Some | grep -v repr
```

**Comment by tcharding** [comment]
Related issue: #1638

**Comment by tcharding** [comment]
These abstractions are non-trivial to reason about and review - consider doing each type as a separate PR.

**Comment by tcharding** [comment]
Related issue: https://github.com/rust-bitcoin/rust-bitcoin/issues/1632

**Comment by PaarthAgarwal** [comment]
I'd like to work on this issue

**Comment by Kixunil** [comment]
@PaarthAgarwal go ahead!

**Comment by PaarthAgarwal** [comment]
@Kixunil, my understanding of this issue is that some operations like mul are not feasible for some data types and need to be checked and deleted. Can you explain this issue more precisely? Also, should I look for them in the whole codebase or specific files, and is there any example for this?

**Comment by tcharding** [comment]
- Consider also doing ops in both directions i.e., `Mul<Foo> for Bar` and `Mul<Bar> for foo`.

**Comment by tcharding** [comment]
Question to the team: should we have checked/unchecked/wrapping etc?

**Comment by apoelstra** [comment]
I suspect we should always have a checked variant and a "panic on overflow" variant. Wrapping probably not.

**Comment by yancyribbens** [comment]
I think this issue confilcts with the conversation on https://github.com/rust-bitcoin/rust-bitcoin/issues/2086.  This issue is older from May 20, while #2086 was last discussed this September.

Specifically the two comments by @apoelstra conflict I think
> I suspect we should always have a checked variant and a "panic on overflow" variant. Wrapping probably not.

and from #2086
>  I think that if "panic in debug mode, overflow in release" is good enough for stdlib it should be good enough for us.

Am I missing some nuance to the conversations?

**Comment by yancyribbens** [comment]
In short, I feel like there's some confusion (in my mind) about how we want to handle wrapping/saturating vs just plain painc.

**Comment by apoelstra** [comment]
@tcharding maybe it'd be fun to write a script for rust-secp (which has #1880) that looks for `Add` without `AddAssign` and same for `Mul`, `Div`, `Sub`. I think that's a single unambigous improvement.

Though maybe we should first check the master list of clippy lints to see if clippy can do this for us.

@yancyribbens I don't see the conflict between those two comments. We should have both a `Result` variant and a "panic" variant. I guess you are unsure whether we should always panic or copy stdlib "panic in debug mode, overflow in release"? I don't feel strongly about that.

**Comment by Kixunil** [comment]
Anyone writing the check needs to be careful to consider that the return type might not be `Self` in which case `AddAssign` is impossible.

**Comment by yancyribbens** [comment]
> I guess you are unsure whether we should always panic or copy stdlib "panic in debug mode, overflow in release"?

Yes, I guess I'm wondering if we should always panic in debug mode and overflow in release mode.  I guess the consensus is we maintain checked variants which have the same behavior in both debug and release mode, and otherwise if it's not a checked variant, it has the same behavior as stdlib which makes sense.   My question was more do we want to maintain all the checked variants in perpetuity instead of making library authors do that work, which I guess is yes.

**Comment by yancyribbens** [comment]
> Anyone writing the check needs to be careful to consider that the return type might not be Self in which case AddAssign is impossible.

I'm not sure what is meant by this.

**Comment by apoelstra** [comment]
>Yes, I guess I'm wondering if we should always panic in debug mode and overflow 

If we introduce explicit panics they should happen in both debug and release mode. If we "inherit" the panics by just using arithmetic ops on primitive numbers then we'll leave the "panic only in debug mode" behavior be, since it's annoying and extra work to make them unconditional. In our docs we can say that we'll panic in debug mode and "might panic" in release mode.

At least, that's my vote.

>I'm not sure what is meant by this.

If you add two objects of type A and the result is of type B, then `a += a` has no meaning and you can't impl `AddAssign`.

My feeling is that this is rare enough that we could get away without the check, at least at first, but I might be wrong.

**Comment by Kixunil** [comment]
> My feeling is that this is rare enough that we could get away without the check, at least at first, but I might be wrong.

I know three cases by heart that we already have. `FeeRate` * `Weight` = `Amount`, `Amount` / `FeeRate` = `Weight`, and `Amount` / `Weight` = `FeeRate`.

So unless the tool has some kind of opt-out for specific cases it'll fail immediately.

**Comment by apoelstra** [comment]
Oh, yeah, for `Mul` it'll be super common. Good point.

**Comment by tcharding** [comment]
> since it's annoying and extra work to make them unconditional

I hit another annoyance the other day. Types that have an invariant that they are less than some MAX are complicated to implement `ops` traits for and maintain the "panic in debug overflow in release" logic. I elected to just not implement `ops` traits and just add `checked_` and `saturating_` functions, forcing users to make the decision. (This was while doing stuff with the locktime `Height` types.)

**Comment by apoelstra** [comment]
> Types that have an invariant that they are less than some MAX are complicated to implement `ops` traits for and maintain the "panic in debug overflow in release" logic.

I think we should just unconditionally panic in these cases.

**Comment by tcharding** [comment]
FTR, in general, I do not understand exactly why its ok to unconditionally panic for input that comes from outside the lib - isn't that an attack vector? Doesn't need answering now, I just wanted to note it.

**Comment by Kixunil** [comment]
@tcharding it is believed that anything else is impractical, although I do like monadic overflow handling, especially for `bitcoin`&co, if we don't account for it not being usual Rust (but I don't think it's unidiomatic).

**Comment by apoelstra** [comment]
@tcharding frankly, it's "ok" because `std` already does it so users already need to audit their entire codebase for invisible DoS vectors.

**Comment by Sh0g0-1758** [comment]
@tcharding, are you working on this issue?

**Comment by tcharding** [comment]
I am not currently working on this issue.

**Comment by Sh0g0-1758** [comment]
Oh, okay. I will have a go at it then.

**Comment by Sh0g0-1758** [comment]
@tcharding , I have raised a PR for this issue, but I don't understand what you meant by this: 

> * Consider also doing ops in both directions i.e., `Mul<Foo> for Bar` and `Mul<Bar> for foo`.

**Comment by tcharding** [comment]
The generic is on the right hand side of the equation X * Y so if we want to be able to write Y * X then two impls of `Mul` are required.

**Comment by tcharding** [comment]
This issue requires a fair bit of thought and also a fair bit of review - I do not believe it is a good first issue.
